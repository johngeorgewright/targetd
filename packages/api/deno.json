{
  "name": "@targetd/api",
  "version": "7.1.0",
  "exports": "./src/index.ts",
  "fmt": {
    "semiColons": false,
    "singleQuote": true
  },
  "tasks": {
    "pwd": "echo $(pwd)",
    "test": "deno test --allow-all --trace-leaks"
  },
  "imports": {
    "@myty/promise-chain": "jsr:@myty/promise-chain@^0.4.6",
    "fast-deep-equal": "npm:fast-deep-equal@^3.0.0",
    "ts-toolbelt": "npm:ts-toolbelt@^9.0.0",
    "zod": "npm:zod@^4.0.15"
  },
  "publish": {
    "include": [
      "deno.json",
      "src/",
      "LICENSE",
      "README.md"
    ]
  }
}
