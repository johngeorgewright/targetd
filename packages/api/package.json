{
  "name": "@targetd/api",
  "version": "6.0.1",
  "description": "Load and query data",
  "main": "dist/index.js",
  "sideEffects": false,
  "scripts": {
    "build": "yarn clean && yarn tsc",
    "clean": "yarn rimraf dist",
    "start": "yarn tsc --watch --preserveWatchOutput",
    "test": "yarn jest --passWithNoTests"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/johngeorgewright/config/issues"
  },
  "homepage": "https://github.com/johngeorgewright/config#readme",
  "devDependencies": {
    "@semantic-release/commit-analyzer": "11.1.0",
    "@semantic-release/exec": "6.0.3",
    "@semantic-release/git": "10.0.1",
    "@semantic-release/github": "9.2.6",
    "@semantic-release/release-notes-generator": "12.1.0",
    "@types/jest": "29.5.13",
    "@types/node": "20.16.9",
    "jest": "29.7.0",
    "prettier-2": "npm:prettier@^2",
    "rimraf": "6.0.1",
    "semantic-release": "23.1.1",
    "semantic-release-monorepo": "8.0.2",
    "ts-jest": "29.2.5",
    "typescript": "5.4.5",
    "zod": "3.23.8"
  },
  "dependencies": {
    "fast-deep-equal": "3.1.3",
    "ts-toolbelt": "9.6.0",
    "tslib": "2.7.0"
  },
  "peerDependencies": {
    "zod": "^3.20.0"
  }
}
