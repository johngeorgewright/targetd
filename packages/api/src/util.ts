import type { L } from 'ts-toolbelt'
import type { MaybePromise } from './types.ts'

export function objectMap<O extends Record<string, unknown>, R>(
  obj: O,
  fn: <K extends keyof O>(v: O[K], k: K) => R,
): Record<keyof O, R> {
  return objectEntries(obj).reduce(
    (result, [key, value]) => ({ ...result, [key]: fn(value, key) }),
    {} as Record<keyof O, R>,
  )
}

export function objectKeys<O extends Record<string, unknown>>(
  obj: O,
): (keyof O)[] {
  return Object.keys(obj)
}

export function objectSize(obj: Record<string, unknown>) {
  return Object.keys(obj).length
}

export function objectEntries<T extends Record<string, unknown>>(obj: T) {
  return Object.entries(obj) as Entries<T>[]
}

export function omit<T extends Record<string, unknown>, K extends keyof T>(
  obj: T,
  keys: K[],
): Omit<T, K> {
  return keys.reduce(
    (obj, key) => {
      delete obj[key]
      return obj
    },
    { ...obj },
  )
}

export function* objectIterator<T extends Record<string, unknown>>(
  obj: T,
): Generator<Entries<Required<T>>> {
  for (const key in obj) {
    if (Object.prototype.hasOwnProperty.call(obj, key)) yield [key, obj[key]]
  }
}

export function objectFitler<T extends Record<string, unknown>>(
  obj: T,
  predicate: <K extends keyof T>(value: T[K], key: K) => boolean,
): Partial<T> {
  const result: Partial<T> = {}

  for (const [key, value] of objectIterator(obj)) {
    if (predicate(value, key)) result[key] = value
  }

  return result
}

export function objectSome<T extends Record<string, unknown>>(
  obj: T,
  predicate: <K extends keyof T>(value: T[K], key: K) => boolean,
): boolean {
  for (const [key, value] of objectIterator(obj)) {
    if (predicate(value, key)) return true
  }

  return false
}

export function someKeysIntersect(
  aObj: Record<string, unknown>,
  bObj: Record<string, unknown>,
) {
  return objectSome(aObj, (_, key) => key in bObj)
}

export function intersection<T extends Record<string, unknown>>(
  aObj: T,
  bObj: Record<string, unknown>,
) {
  return objectFitler(aObj, (_, key) => key in bObj)
}

export function intersectionKeys<T extends Record<string, unknown>>(
  aObj: T,
  bObj: Record<string, unknown>,
): (keyof Partial<T>)[] {
  const keys: (keyof Partial<T>)[] = []
  for (const [key] of objectIterator(aObj)) if (key in bObj) keys.push(key)
  return keys
}

class EveryAsyncFail extends Error {}

/**
 * Call an asynchronous function on all key/value pairs.
 * Returns true only if all calls return true.
 *
 * @example
 * if (await objectEveryAsync({ foo: 'bar' }, (value) => Promise.resolve(value === 'bar'))) {
 *   console.info('Everything is bar')
 * }
 */
export async function objectEveryAsync<T extends Record<string, unknown>>(
  obj: T,
  fn: <K extends keyof T>(value: T[K], key: K) => MaybePromise<boolean>,
) {
  try {
    await Promise.all(
      objectEntries(obj).map(async ([key, value]) => {
        if (!(await Promise.resolve(fn(value, key)))) throw new EveryAsyncFail()
      }),
    )
  } catch (error) {
    if (error instanceof EveryAsyncFail) return false
    else throw error
  }
  return true
}

export function arrayInit<T extends unknown[]>(array: T) {
  return array.slice(0, -1) as L.Pop<T>
}

export function arrayLast<T extends unknown[]>(array: T): L.Last<T> {
  return array[array.length - 1]
}

type Entries<T extends Record<string | symbol, unknown>> = {
  [K in keyof T]: [K, T[K]]
}[keyof T]
