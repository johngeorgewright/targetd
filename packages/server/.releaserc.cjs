// @ts-check

/**
 * @type {import('npm:semantic-release').Options}
 */
module.exports = {
  extends: [
    '../../.releaserc.cjs',
  ],
}
