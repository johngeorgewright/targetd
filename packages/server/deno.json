{
  "name": "@targetd/server",
  "version": "6.0.0",
  "exports": "./src/index.ts",
  "imports": {
    "@johngw/error": "npm:@johngw/error@^2.2.2",
    "@plugola/query-params": "npm:@plugola/query-params@^2.0.2",
    "cors": "npm:cors@^2.8.5",
    "express": "npm:express@^4.18.3",
    "express-query-boolean": "npm:express-query-boolean@^2.0.0",
    "query-types": "npm:query-types@^0.1.4",
    "zod": "npm:zod@^3.25.0"
  },
  "fmt": {
    "semiColons": false,
    "singleQuote": true
  },
  "tasks": {
    "test": "deno test --allow-all --trace-leaks"
  }
}
