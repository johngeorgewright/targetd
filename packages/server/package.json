{
  "name": "@targetd/server",
  "version": "6.0.0",
  "description": "HTTP server for targetd data",
  "main": "dist/index.js",
  "sideEffects": false,
  "scripts": {
    "build": "yarn clean && yarn tsc",
    "clean": "yarn rimraf dist",
    "start": "yarn tsc --watch --preserveWatchOutput",
    "release": "yarn semantic-release -e semantic-release-monorepo",
    "test": "yarn jest --passWithNoTests"
  },
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/johngeorgewright/config/issues"
  },
  "homepage": "https://github.com/johngeorgewright/config#readme",
  "devDependencies": {
    "@semantic-release/commit-analyzer": "11.1.0",
    "@semantic-release/exec": "6.0.3",
    "@semantic-release/git": "10.0.1",
    "@semantic-release/github": "9.2.6",
    "@semantic-release/release-notes-generator": "12.1.0",
    "@targetd/api": "workspace:*",
    "@targetd/date-range": "workspace:*",
    "@types/cors": "2.8.17",
    "@types/jest": "29.5.13",
    "@types/lodash": "^4.14.202",
    "@types/node": "20.16.9",
    "@types/supertest": "6.0.2",
    "jest": "29.7.0",
    "lodash": "4.17.21",
    "rimraf": "6.0.1",
    "semantic-release": "23.1.1",
    "semantic-release-monorepo": "8.0.2",
    "supertest": "6.3.4",
    "ts-jest": "29.2.5",
    "typescript": "5.4.5",
    "zod": "3.23.8"
  },
  "dependencies": {
    "@johngw/error": "2.2.2",
    "@plugola/query-params": "2.0.2",
    "@types/express": "4.17.21",
    "@types/express-serve-static-core": "4.19.6",
    "cors": "2.8.5",
    "express": "4.18.3",
    "express-query-boolean": "2.0.0",
    "query-types": "0.1.4",
    "tslib": "2.7.0"
  },
  "peerDependencies": {
    "@targetd/api": ">=6.0.1",
    "zod": "^3.20.0"
  }
}
