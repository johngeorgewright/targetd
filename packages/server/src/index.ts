import type { DT } from '@targetd/api'
import cors from 'cors'
// @ts-types='npm:@types/express@4'
import express from 'express'
import queryTypes from 'query-types'
import { errorHandler } from './middleware/error.ts'
import { StatusError } from './StatusError.ts'
import { castQueryArrayProps } from './middleware/castQueryArrayProps.ts'

/**
 * @param data
 * @param pathStructure Use a path structure when you want to create a route that uses request params
 */
export function createServer<
  D extends DT.Any,
  App extends express.Express = express.Express,
>(
  data: D | (() => D),
  {
    app = express() as App,
    pathStructure,
  }: {
    app?: App
    pathStructure?: (keyof DT.QueryParsers<D>)[]
  } = {},
): App {
  const getData = typeof data === 'function' ? data : () => data

  let server = app.use(cors())

  if (pathStructure) {
    server = server.get(
      `/:${pathStructure.join('/:')}`,
      queryTypes.middleware(),
      castQueryArrayProps(getData),
      async (req, res, next) => {
        let payloads: Record<string, any>
        try {
          payloads = await getData().getPayloadForEachName({
            ...req.params,
            ...req.query,
          })
        } catch (err) {
          return next(err)
        }
        res.json(payloads)
      },
    )
  }

  return server
    .get(
      '/:name',
      queryTypes.middleware(),
      castQueryArrayProps(getData),
      async (req, res, next) => {
        const data = getData()

        if (!(req.params.name in data.payloadParsers)) {
          return next(
            new StatusError(404, `Unknown data property ${req.params.name}`),
          )
        }

        let payload: any
        try {
          payload = await data.getPayload(req.params.name, req.query as any)
        } catch (err) {
          return next(err)
        }

        if (payload === undefined) res.sendStatus(204)
        else res.json(payload)
      },
    )
    .get(
      '/',
      queryTypes.middleware(),
      castQueryArrayProps(getData),
      async (req, res, next) => {
        let payloads: Record<string, any>
        try {
          payloads = await getData().getPayloadForEachName(req.query as any)
        } catch (err) {
          return next(err)
        }
        res.json(payloads)
      },
    )
    .use(errorHandler())
}
