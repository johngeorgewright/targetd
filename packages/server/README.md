# @targetd/server

> HTTP server for targetd data

## Installation

| JS Runtime | Command                                     |
| ---------- | ------------------------------------------- |
| Node.js    | `npx jsr add @targetd/api @targetd/server`  |
| Bun        | `bunx jsr add @targetd/api @targetd/server` |
| Deno       | `deno add @targetd/api @targetd/server`     |

## Exammple

```typescript
import { Data } from '@targetd/api',
import { createServer } from '@targetd/server'
import z from 'zod'

const data = Data.create({
  data: {
    foo: z.string(),
    b: z.string()
  }
})

createServer(data).listen(8080)
```
