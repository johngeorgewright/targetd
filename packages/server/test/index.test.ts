import { assertSnapshot } from 'jsr:@std/testing/snapshot'
import { Data, targetEquals, targetIncludes } from '@targetd/api'
import dateRangeTargeting from '@targetd/date-range'
import _ from 'npm:lodash'
import { setTimeout } from 'node:timers/promises'
// @ts-types='npm:@types/supertest'
import request from 'npm:supertest'
import z from 'zod/v4'
import { createServer } from '@targetd/server'
import { promisify } from 'node:util'

Deno.test('get one data point', async () => {
  await using disposable = await createDisposableServer()
  const { server } = disposable

  await request(server)
    .get('/foo')
    .expect('Content-Type', /json/)
    .expect(200)
    .expect('"bar"')

  await request(server)
    .get('/foo?weather=sunny')
    .expect('Content-Type', /json/)
    .expect(200)
    .expect('"😎"')

  await request(server)
    .get('/foo?weather=rainy')
    .expect('Content-Type', /json/)
    .expect(200)
    .expect('"☂️"')

  await request(server)
    .get('/foo?highTide=true')
    .expect('Content-Type', /json/)
    .expect(200)
    .expect('"🌊"')

  await request(server)
    .get('/foo?highTide=true&weather=sunny')
    .expect('Content-Type', /json/)
    .expect(200)
    .expect('"🏄‍♂️"')

  await request(server)
    .get('/foo?asyncThing=true')
    .expect('Content-Type', /json/)
    .expect(200)
    .expect('"Async payload"')

  await request(server)
    .get('/timed?date[start]=2002-01-01')
    .expect('Content-Type', /json/)
    .expect(200)
    .expect('"in time"')

  await request(server)
    .get('/timed?date[start]=2012-01-01')
    .expect('Content-Type', /json/)
    .expect(200)
    .expect('"out of time"')

  await request(server)
    .get('/foo?arrayThing=a')
    .expect('Content-Type', /json/)
    .expect(200)
    .expect('"a t\'ing"')

  await request(server)
    .get('/foo?arrayThing=a&arrayThing=b')
    .expect('Content-Type', /json/)
    .expect(200)
    .expect('"b t\'ing"')
})

Deno.test('get all', async (t) => {
  await using disposable = await createDisposableServer()
  const { server } = disposable

  let response = await request(server)
    .get('/')
    .expect('Content-Type', /json/)
    .expect(200)
  await assertSnapshot(t, response.body)

  response = await request(server)
    .get('/?weather=sunny')
    .expect('Content-Type', /json/)
    .expect(200)
  await assertSnapshot(t, response.body)

  response = await request(server)
    .get('/?weather=rainy')
    .expect('Content-Type', /json/)
    .expect(200)
  await assertSnapshot(t, response.body)

  response = await request(server)
    .get('/?highTide=true')
    .expect('Content-Type', /json/)
    .expect(200)
  await assertSnapshot(t, response.body)

  response = await request(server)
    .get('/?highTide=true&weather=sunny')
    .expect('Content-Type', /json/)
    .expect(200)
  await assertSnapshot(t, response.body)

  response = await request(server)
    .get('/?asyncThing=true')
    .expect('Content-Type', /json/)
    .expect(200)
  await assertSnapshot(t, response.body)
})

async function createDisposableServer() {
  const app = createServer(await createData())
  const { promise, resolve } = Promise.withResolvers<void>()
  const server = app.listen(0, resolve)
  await promise
  return {
    server,
    [Symbol.asyncDispose]: promisify(server.close.bind(server)),
  }
}

async function createData() {
  return await Data.create()
    .usePayload({
      foo: z.string(),
      bar: z.number(),
      timed: z.string(),
    })
    .useTargeting({
      weather: targetIncludes(z.string()),
      highTide: targetEquals(z.boolean()),
      asyncThing: {
        predicate: (q) =>
          setTimeout(10, (t) => q === t && setTimeout(10, true)),
        queryParser: z.boolean(),
        targetingParser: z.boolean(),
      },
      arrayThing: {
        predicate: (q) => (t) => _.difference(q, t).length === 0,
        queryParser: z.string().array(),
        targetingParser: z.string().array(),
      },
      date: dateRangeTargeting,
    })
    .addRules('foo', [
      {
        targeting: {
          highTide: true,
          weather: ['sunny'],
        },
        payload: '🏄‍♂️',
      },
      {
        targeting: {
          weather: ['sunny'],
        },
        payload: '😎',
      },
      {
        targeting: {
          weather: ['rainy'],
        },
        payload: '☂️',
      },
      {
        targeting: {
          highTide: true,
        },
        payload: '🌊',
      },
      {
        targeting: {
          asyncThing: true,
        },
        payload: 'Async payload',
      },
      {
        targeting: {
          arrayThing: ['a'],
        },
        payload: "a t'ing",
      },
      {
        targeting: {
          arrayThing: ['a', 'b'],
        },
        payload: "b t'ing",
      },
      {
        payload: 'bar',
      },
    ])
    .addRules('bar', [
      {
        payload: 123,
      },
    ])
    .addRules('timed', [
      {
        targeting: {
          date: { start: '2001-01-01', end: '2010-01-01' },
        },
        payload: 'in time',
      },
      {
        payload: 'out of time',
      },
    ])
}
