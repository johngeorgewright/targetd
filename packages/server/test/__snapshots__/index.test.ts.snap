export const snapshot = {};

snapshot[`get all 1`] = `
{
  bar: 123,
  foo: "bar",
  timed: "out of time",
}
`;

snapshot[`get all 2`] = `
{
  bar: 123,
  foo: "😎",
  timed: "out of time",
}
`;

snapshot[`get all 3`] = `
{
  bar: 123,
  foo: "☂️",
  timed: "out of time",
}
`;

snapshot[`get all 4`] = `
{
  bar: 123,
  foo: "🌊",
  timed: "out of time",
}
`;

snapshot[`get all 5`] = `
{
  bar: 123,
  foo: "🏄‍♂️",
  timed: "out of time",
}
`;

snapshot[`get all 6`] = `
{
  bar: 123,
  foo: "Async payload",
  timed: "out of time",
}
`;
